### PR TITLE
Expand dns1123 processing

### PR DESCRIPTION
**Pull request by jaysonesmith**
Tests show various formats that get processed as well as some standard success cases.

My idea around the code is to only perform the transformations that are actually needed for a particular input along with a fallback.

Before and after performance numbers:

```
Running tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench ^(BenchmarkMakeDNS1123Compatible|BenchmarkMakeDNS1123CompatibleOld)$ github.com/freshly/tuber/pkg/reviewapps

goos: darwin
goarch: amd64
pkg: github.com/freshly/tuber/pkg/reviewapps
cpu: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
BDNS1123Compatible/only_numbers-12                                 4723400	      254.9 ns/op	      16 B/op	       2 allocs/op
BDNS1123Compatible/only_lowercase_letters_and_hyphens-12           3231021	      384.5 ns/op	      32 B/op	       2 allocs/op
BDNS1123Compatible/capital_letters-12                              3191713	      382.7 ns/op	      16 B/op	       3 allocs/op
BDNS1123Compatible/underscores_to_hyphens-12                       1425891	      830.4 ns/op	      48 B/op	       3 allocs/op
BDNS1123Compatible/capitals_and_underscores-12                     1171012	      1010 ns/op	      64 B/op	       4 allocs/op
BDNS1123Compatible/too_long_only_alphanumeric-12                   1607484	      754.9 ns/op	     112 B/op	       2 allocs/op
BDNS1123Compatible/symbols-12                                       684150	      1867 ns/op	      72 B/op	       5 allocs/op
BDNS1123Compatible/symbols_without_modifying_valid_hyphens-12       597316	      1840 ns/op	      72 B/op	       5 allocs/op
BDNS1123Compatible/leading_hyphens-12                              1632974	      690.8 ns/op	      40 B/op	       3 allocs/op
BDNS1123Compatible/trailing_hyphens-12                             1279946	      931.8 ns/op	      40 B/op	       3 allocs/op
BDNS1123Compatible/only_hyphens-12                                 1742583	      726.5 ns/op	      65 B/op	       4 allocs/op
BDNS1123CompatibleOld/only_numbers-12                               363850	      2916 ns/op	    2371 B/op	      35 allocs/op
BDNS1123CompatibleOld/only_lowercase_letters_and_hyphens-12         405228	      3170 ns/op	    2403 B/op	      35 allocs/op
BDNS1123CompatibleOld/capital_letters-12                            395342	      3618 ns/op	    2371 B/op	      36 allocs/op
BDNS1123CompatibleOld/underscores_to_hyphens-12                     279961	      3595 ns/op	    2372 B/op	      35 allocs/op
BDNS1123CompatibleOld/capitals_and_underscores-12                   350478	      3569 ns/op	    2385 B/op	      36 allocs/op
BDNS1123CompatibleOld/too_long_only_alphanumeric-12                 270283	      4493 ns/op	    2657 B/op	      35 allocs/op
BDNS1123CompatibleOld/symbols-12                                    326035	      3705 ns/op	    2410 B/op	      36 allocs/op
BDNS1123CompatibleOld/symbols_without_modifying_valid_hyphens-12    345802	      3535 ns/op	    2412 B/op	      36 allocs/op
BDNS1123CompatibleOld/leading_hyphens-12                            316065	      3214 ns/op	    2368 B/op	      35 allocs/op
BDNS1123CompatibleOld/trailing_hyphens-12                           422691	      2943 ns/op	    2371 B/op	      35 allocs/op
BDNS1123CompatibleOld/only_hyphens-12                               410724	      2903 ns/op	    2332 B/op	      32 allocs/op
PASS
ok  	github.com/freshly/tuber/pkg/reviewapps	32.916s
```

### CODE REVIEW

File: pkg/reviewapps/reviewapps_test.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+		},
+		{
+			name:     "capitals and underscores",
+			input:    "A_B_C_D_E",
+			expected: "a-b-c-d-e",
+		},
+		{
+			name:     "too long only alphanumeric",
+			input:    strings.Repeat("a", dns1123NameMaximumLength+1),
+			expected: strings.Repeat("a", dns1123NameMaximumLength),
```
Comment (quinn-freshly) [inline pkg/reviewapps/reviewapps_test.go:46]: maybe i'm old fashioned, but its nice to see the explicit value and the explicit output in tests. This doesn't make the test resilient to changes in the "max length" constant, but i prefer to see the failing test after the constant is changed, then updated to be green again.

File: pkg/reviewapps/reviewapps_test.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+		},
+		{
+			name:     "trailing hyphen",
+			input:    "foo-",
+			expected: "foo",
+		},
+		{
+			name:     "only hyphens",
+			input:    "----",
+			expected: fmt.Sprintf("review-app-%d", time.Now().Unix()),
```
Comment (quinn-freshly) [inline pkg/reviewapps/reviewapps_test.go:66]: nice.

File: pkg/reviewapps/reviewapps_test.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+			expected: "foobarfoobar",
+		},
+		{
+			name:     "symbols without modifying valid hyphens",
+			input:    "foo&bar-foo@bar",
+			expected: "foobar-foobar",
+		},
+		{
+			name:     "trailing hyphen",
+			input:    "foo-",
```
Comment (quinn-freshly) [inline pkg/reviewapps/reviewapps_test.go:60]: can you add a double trailing hyphen here, or add a test case for more than one hyphen?

**Review by quinn-freshly — CHANGES_REQUESTED** [review]
this is great, thanks for adding test cases for this. a few changes i think would be nice, but you can disregard. 

Also, I wonder if it would be worth adding in the hyphen plus 4 or 5 random alphanum characters at the end inside this PR?

File: pkg/reviewapps/reviewapps_test.go (hunk 3 above)
Comment (jaysonesmith) [reply pkg/reviewapps/reviewapps_test.go:60]: Yea of course!

File: pkg/reviewapps/reviewapps_test.go (hunk 2 above)
Comment (jaysonesmith) [reply pkg/reviewapps/reviewapps_test.go:66]: 🎉  Normally finding a way to inject a custom time object for testing but don't want to change the signature.

File: pkg/reviewapps/reviewapps_test.go (hunk 1 above)
Comment (jaysonesmith) [reply pkg/reviewapps/reviewapps_test.go:46]: Yea, totally fine.

Comment (jaysonesmith) [reply pkg/reviewapps/reviewapps_test.go:46]: I'm updating the test case to have hard coded values with the input being > 63 characters as that's what the actual spec dictates. This'll mean we'll only need to change the expected value if we do end up changing the const value down the line.

**Comment by jaysonesmith** [comment]
@quinn-freshly Good call on the leading hyphens as that actually wasn't handled. 🙏

**Review by quinn-freshly — APPROVED** [review]

File: pkg/reviewapps/reviewapps.go [hunk 4]
```diff
@@ -174,28 +176,46 @@ func DeleteReviewApp(ctx context.Context, db *core.DB, reviewAppName string, cre
...
+func makeDNS1123Compatible(name string) string {
+	n := []byte(name)
 
-	if len(name) > DNS1123NameMaximumLength {
-		name = name[0:DNS1123NameMaximumLength]
+	if len(n) > dns1123NameMaximumLength {
+		n = n[0:dns1123NameMaximumLength]
+	}
+	for !dns1123LabelRe.Match(n) {
+		switch {
```
Comment (jkminneti) [inline pkg/reviewapps/reviewapps.go:204]: this seems a little unnecessarily complicated
```go
if dns1123LabelRe.Match(n) {
	return string(n)
}
n = bytes.ToLower(n)
n = bytes.ReplaceAll(n, []byte("_"), []byte("-"))
n = symbolsRe.ReplaceAll(n, []byte(""))
n = bytes.Trim(n, "-")
if !dns1123LabelRe.Match(n) {
	return fmt.Sprintf("review-app-%d", time.Now().Unix())
}
```
it's true, as-is, some of those transformations are skipped if unnecessary. but is that worth the extra and repetitive checks, if it's also arguably less readable?

def a style question tho

**Review by jkminneti — APPROVED** [review]
point about style, otherwise lgtm

**Review by Hbbb — APPROVED** [review]
Let's merge this!

File: pkg/reviewapps/reviewapps.go [hunk 5]
```diff
@@ -174,28 +176,46 @@ func DeleteReviewApp(ctx context.Context, db *core.DB, reviewAppName string, cre
...
+		case capitalsRe.Match(n):
+			n = bytes.ToLower(n)
+		case bytes.Contains(n, []byte("_")):
+			n = bytes.ReplaceAll(n, []byte("_"), []byte("-"))
+		case symbolsRe.Match(n):
+			n = symbolsRe.ReplaceAll(n, []byte(""))
+		case bytes.Compare(n, []byte("-")) == 1:
+			n = bytes.Trim(n, "-")
+		default:
+			return fmt.Sprintf("review-app-%d", time.Now().Unix())
```
Comment (jkminneti) [inline pkg/reviewapps/reviewapps.go:214]: turns out it has to end in an alpha character

**Review by jkminneti — CHANGES_REQUESTED** [review]

File: pkg/events/processor_test.go [hunk 6]
```diff
@@ -0,0 +1 @@
+package events
```
Comment (jkminneti) [inline pkg/events/processor_test.go:1]: this supposed to be here?

File: go.sum [hunk 7]
```diff
@@ -23,12 +23,9 @@ cloud.google.com/go/bigquery v1.3.0/go.mod h1:PjpwJnslEMmckchkHFfq+HTD2DmtT67aNF
 cloud.google.com/go/bigquery v1.4.0/go.mod h1:S8dzgnTigyfTmLBfrtrhyYhwRxG72rYxvftPBK2Dvzc=
 cloud.google.com/go/bigquery v1.5.0/go.mod h1:snEHRnqQbz117VIFhE8bmtwIDY80NLUZUMb4Nv6dBIg=
 cloud.google.com/go/bigquery v1.7.0/go.mod h1://okPTzCYNXSlb24MZs83e2Do+h+VXtc4gLoIoXIAPc=
-cloud.google.com/go/bigquery v1.8.0 h1:PQcPefKFdaIzjQFbiyOgAqyx8q5djaE7x9Sqe712DPA=
```
Comment (jkminneti) [inline go.sum:26]: wow i thought this was cus of outdated main, but nope main is untidied

**Review by jkminneti — APPROVED** [review]

File: pkg/events/processor_test.go (hunk 6 above)
Comment (jaysonesmith) [reply pkg/events/processor_test.go]: Nope

**Pull request merged**